pgwire: error cleanly when a simple query is too big

otan opened the pull request:
Refs: #50330

Release justification: low risk, high benefit changes to
existing functionality

Release note (sql change): Previously, certain SQL commands
sent over the PostgreSQL FE/BE protocol to which were too big would
error opaquely. This is now resolved for non-PREPARE related statements,
clearly error messaging instead.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53790)
<!-- Reviewable:end -->

knz commented on pkg/sql/pgwire/pgwirebase/encoding.go:
```diff
@@ -150,18 +155,49 @@ func (b *ReadBuffer) ReadUntypedMsg(rd io.Reader) (int, error) {
+				humanize.IBytes(uint64(size)),
+				humanize.IBytes(uint64(b.maxMessageSize)),
+			),
+			"the maximum message size can be configured using the %s CLUSTER SETTING",
```
cluster setting in lowercase

knz commented on pkg/sql/pgwire/conn.go:
```diff
@@ -321,6 +321,44 @@ Loop:
+			// the todo message by jordan regarding error in extended protocol state
+			// below.
+			if pgwirebase.IsMessageTooBigError(err) && typ == pgwirebase.ClientMsgSimpleQuery {
+				log.VEventf(ctx, 1, "pgwire: found big error message; attempting to slurp bytes and return error: %s", err)
```
The use of `VEventf` is not really warranted here and below -- we don't typically collect nor look at pgwire traces.

I think you can go for `log.Infof` under `log.V(1)`

(if you find yourself wishing for `if log.V(1) { log.Infof() } ` to be a thing without the tracing cruft, feel free to add the wrapper in the `log` package)

otan replied:
Done

knz commented on pkg/sql/pgwire/conn.go:
Can you say more words here in the comment about how this effects the stream of messages: is the client able to recover its session no matter which packet is too large?

or is the tolerance for large packets limited to a certain message type?

I don't see this distinction made in the test:

1. if the code is meant to only tolerate a particular message type, you want a test that demonstrates that the connection is still closed for message types that are not supported

2. conversely, if the code is meant to tolerate excess type _any_ msg type, we want tests for many of them.

All in all I would recommend various combinations of tests in `pgwire_test.go` (i.e. datadriven-tests) with a small message size limit so that you can exercise this in many combinations of message types.

otan replied:
> Can you say more words here in the comment about how this effects the stream of messages: is the client able to recover its session no matter which packet is too large?
> or is the tolerance for large packets limited to a certain message type?

Done.

knz commented on pkg/sql/conn_executor_test.go:
```diff
@@ -924,3 +924,117 @@ func (f *dynamicRequestFilter) filter(
 	return nil
 }
+
+func TestBigClientMessage(t *testing.T) {
```
see my comment below: I believe this test is best served by datadriven testing via pgwire_test.go

otan replied:
Going to push back on this one -- this introduces a lot of cruft with pg wire tests, e.g.:
* needing to setup a new connection for the cluster setting to kick in
* needing to large strings with replace directives
* needed to differentiate hard errors which involve a connection termination (i.e. io.EOF), as opposed to ErrorResponse with retries (as from experience, CLUSTER SETTING in tests may not immediately kick in).
* big messages are unwieldly to commit as an expected result (we currently expect at least 16kB as the shortest buffer size...).
* it does not integration test well with an driver

I've added some more cases here to test through.

knz commented on pkg/sql/pgwire/pgwirebase/too_big_error.go:
```diff
@@ -0,0 +1,88 @@
+// Copyright 2020 The Cockroach Authors.
```
I am impressed by how much effort went into the code in this file!
I will make comments about your implementation below, for this and the next time you want to implement a custom error.

However, meanwhile I believe you don't actually need this machinery for this PR. I don't even think you need a custom error type. A pgwire datadriven test will tell your test everything it needs to know.

otan replied:
I do want one; I need to differentiate errors between:

```
readConn, err := io.ReadAll(...)
// error during streaming here
```

as opposed to the case where we encountered a big size.

knz commented on pkg/sql/pgwire/pgwirebase/too_big_error.go:
```diff
@@ -0,0 +1,88 @@
+
+func (w *withMessageTooBig) Format(s fmt.State, verb rune) { errors.FormatError(w, s, verb) }
+
+func (w *withMessageTooBig) FormatError(p errors.Printer) (next error) {
```
`FormatError` and the `errors.Formatter` interface are not necessary for trivial wrappers -- the formatting code already knows how to deal with those.

knz commented on pkg/sql/pgwire/pgwirebase/too_big_error.go:
```diff
@@ -0,0 +1,88 @@
+		var err error
+		size, err = strconv.Atoi(details[0])
+		if err != nil {
+			return err
```
it's an API violation to return an error which is not the result of decoding.

You have two options:
- either decode fails, in which case you must return `nil` (the errors library will do the right thing and still decode _something_ that's alike to your original object)
- or actually return a `&withMessageTooBig{}`.

otan replied:
ah, done!

knz reviewed (commented):
> Reviewed 2 of 6 files at r1, 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53790)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @otan, and @solongordon)

___
*[pkg/sql/pgwire/pgwirebase/too_big_error.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53790#-MGAGKOIW7OOq_XwR3-7-r1-1:-MGi8hvu6Wf09hqpJ6bx:b-utwe04) ([raw file](https://github.com/cockroachdb/cockroach/blob/37daeb16404922bd85eb89b2aa7813616010d182/pkg/sql/pgwire/pgwirebase/too_big_error.go#L1)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I do want one; I need to differentiate errors between:

```
readConn, err := io.ReadAll(...)
// error during streaming here
```

as opposed to the case where we encountered a big size.
</blockquote></details>

I still think that this code here is both extremely well written and not necessary.

- The machinery of the encode/decode logic is only necessary if you want to transport the errors as Go objects between nodes.  This is not the case here, or is it?

- A dedicated wrapper object with a payload (as opposed to `errors.Wrapf` to embed the problematic size in the error string) is needed if you need to retrieve the payload later. Do you?

I agree you want to identify the error. You can do this as follows:

```go
// in global scope
var errTooLarge = errors.New("size too large")

// where the size error occurs
....
         err = errors.Mark(yourErr, errTooLarge)
...


// where you want to detect it
...
      if errors.Is(err, errTooLarge)
...
```
___
*[pkg/util/log/log.go, line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53790#-MGi8ZTY25LG5PrpQGPM:-MGi8ZTY25LG5PrpQGPN:bg60iqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/38e8529f846f4290d95447f12014cada7ee6f78c/pkg/util/log/log.go#L90)):*
> ```Go
> // appended.
> func VInfof(ctx context.Context, level Level, format string, args ...interface{}) {
> 	if V(level) {
> ```

nice :100: 


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/pgwire/pgwirebase/too_big_error.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53790#-MGAGKOIW7OOq_XwR3-7-r1-1:-MGiAemvDIPDoxG7tVt9:blcc1gh) ([raw file](https://github.com/cockroachdb/cockroach/blob/37daeb16404922bd85eb89b2aa7813616010d182/pkg/sql/pgwire/pgwirebase/too_big_error.go#L1)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I still think that this code here is both extremely well written and not necessary.

- The machinery of the encode/decode logic is only necessary if you want to transport the errors as Go objects between nodes.  This is not the case here, or is it?

- A dedicated wrapper object with a payload (as opposed to `errors.Wrapf` to embed the problematic size in the error string) is needed if you need to retrieve the payload later. Do you?

I agree you want to identify the error. You can do this as follows:

```go
// in global scope
var errTooLarge = errors.New("size too large")

// where the size error occurs
....
         err = errors.Mark(yourErr, errTooLarge)
...


// where you want to detect it
...
      if errors.Is(err, errTooLarge)
...
```
</blockquote></details>

Done. Thanks for the tip - didn't know about mark.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/pgwire/pgwirebase/too_big_error.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53790#-MGAGKOIW7OOq_XwR3-7-r1-1:-MGiBz019P9-eIyv5sPp:b4r9yyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/37daeb16404922bd85eb89b2aa7813616010d182/pkg/sql/pgwire/pgwirebase/too_big_error.go#L1)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done. Thanks for the tip - didn't know about mark.
</blockquote></details>

Ah, but I need the size it would exceed by -- is there any way to encode that nicely?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/pgwire/pgwirebase/too_big_error.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53790#-MGAGKOIW7OOq_XwR3-7-r1-1:-MGiDajN1LKyP_-ZB3S0:b8ynfrz) ([raw file](https://github.com/cockroachdb/cockroach/blob/37daeb16404922bd85eb89b2aa7813616010d182/pkg/sql/pgwire/pgwirebase/too_big_error.go#L1)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Ah, but I need the size it would exceed by -- is there any way to encode that nicely?
</blockquote></details>

(this is so that the upper loop can slurp the remaining bytes safely. I think it's neater to do this in the error rather than storing it on the ReadBuffer)


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53790)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @otan, and @solongordon)

___
*[pkg/sql/pgwire/pgwirebase/too_big_error.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53790#-MGAGKOIW7OOq_XwR3-7-r1-1:-MGiE5Yj2GZw1SqahQ6S:b-7bcz2e) ([raw file](https://github.com/cockroachdb/cockroach/blob/37daeb16404922bd85eb89b2aa7813616010d182/pkg/sql/pgwire/pgwirebase/too_big_error.go#L1)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

(this is so that the upper loop can slurp the remaining bytes safely. I think it's neater to do this in the error rather than storing it on the ReadBuffer)
</blockquote></details>

oops I didn't notice that. Well then the wrapper is fine, but you probably don't need to implement `Cause()` nor encode/decode.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53790)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @otan, and @solongordon)

___
*[pkg/sql/pgwire/pgwirebase/too_big_error.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53790#-MGAGKOIW7OOq_XwR3-7-r1-1:-MGiEH2GEJACpNEjwBKz:b-m2ej4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/37daeb16404922bd85eb89b2aa7813616010d182/pkg/sql/pgwire/pgwirebase/too_big_error.go#L1)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

oops I didn't notice that. Well then the wrapper is fine, but you probably don't need to implement `Cause()` nor encode/decode.
</blockquote></details>

(you need either `Cause` _or_ `Unwrap` for our `errors` library; `Cause` is only needed for interop with the legacy package, but since your error is not traveling over the network you don't need it)


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/pgwire/pgwirebase/too_big_error.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53790#-MGAGKOIW7OOq_XwR3-7-r1-1:-MGiFzzI1rOwitp7kb-W:bros63i) ([raw file](https://github.com/cockroachdb/cockroach/blob/37daeb16404922bd85eb89b2aa7813616010d182/pkg/sql/pgwire/pgwirebase/too_big_error.go#L1)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

(you need either `Cause` _or_ `Unwrap` for our `errors` library; `Cause` is only needed for interop with the legacy package, but since your error is not traveling over the network you don't need it)
</blockquote></details>

Cool, thanks for the guidance. And the errors package is a pleasure to use, thank you!


<!-- Sent from Reviewable.io -->

otan commented:
bors r+

otan commented:
bors r-

otan commented:
bors r=knz

otan commented:
bors r-

oops, stressrace

otan committed:
pgwire: error cleanly when a simple query is too big

Release justification: low risk, high benefit changes to
existing functionality

Release note (sql change): Previously, certain SQL commands
sent over the PostgreSQL FE/BE protocol to which were too big would
error opaquely. This is now resolved for non-PREPARE related statements,
clearly error messaging instead.

otan commented:
bors r=knz
